File: lib/AbstractLayer/AbsLayerIdentityManager.swift
//
//  AbsLayerIdentityManager.swift
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

extension AbstractLayer {
  final class IdentityManager {
    private init() {}

    public static func GetMnemonic(language: String, words: String) -> String? {
      let languagePtr = String.ToUnsafeMutablePointer(data: language)
      let wordsPtr = String.ToUnsafeMutablePointer(data: words)
      
      let mnemonicPtr = AbsLayer_IdentityManager_GetMnemonic(languagePtr, wordsPtr)
      let mnemonic = String.FromUnsafeMutablePointer(data: mnemonicPtr)
      AbstractLayer_FreeBuf(mnemonicPtr)
      
      return mnemonic
    }
    
    public static func GetSeed(mnemonic: String, language: String,
                               words: String, mnemonicPassword: String) -> String? {
      let mnemonicPtr = String.ToUnsafeMutablePointer(data: mnemonic)
      let languagePtr = String.ToUnsafeMutablePointer(data: language)
      let wordsPtr = String.ToUnsafeMutablePointer(data: words)
      let mnemonicPasswordPtr = String.ToUnsafeMutablePointer(data: mnemonicPassword)
      
      let seedPtr = AbsLayer_IdentityManager_GetSeed(mnemonicPtr, languagePtr,
                                                          wordsPtr, mnemonicPasswordPtr)
      let seed = String.FromUnsafeMutablePointer(data: seedPtr)
      AbstractLayer_FreeBuf(seedPtr)
      
      return seed
    }
    
    public static func CreateIdentity(localPath: String) -> Int64 {
      let localPathPtr = String.ToUnsafeMutablePointer(data: localPath)
      
      let obj = AbsLayer_IdentityManager_CreateIdentity(localPathPtr)
      
      return obj
    }
    
  }
}

File: lib/AbstractLayer/AbsLayerIdentityManager.cpp
//
//  AbsLayerIdentityManager.cpp
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

#include "AbsLayerIdentityManager.hpp"

#include "IdentityManager.h"
#include "Identity.h"
#include "WalletError.h"

char* AbsLayer_IdentityManager_GetMnemonic(const char* language, const char* words)
{
  std::string mnemonic = elastos::IdentityManager::GetMnemonic(language, words);
  char* mnemonicBuf = strdup(mnemonic.c_str());
  return mnemonicBuf;
}

char* AbsLayer_IdentityManager_GetSeed(const char* mnemonic, const char* mnemonicPassword)
{
  std::string seed = elastos::IdentityManager::GetSeed(mnemonic, mnemonicPassword);
  char* seedBuf = strdup(seed.c_str());
  return seedBuf;
}

int64_t AbsLayer_IdentityManager_CreateIdentity(const char* localPath)
{
  std::shared_ptr<elastos::Identity>* identity = new std::shared_ptr<elastos::Identity>();
  int ret = elastos::IdentityManager::CreateIdentity(localPath, identity);
  if (ret != E_WALLET_C_OK) {
    delete identity;
    return 0;
  }
  
  return reinterpret_cast<int64_t>(identity);

}
